### DISCUSSION TITLE
API/field object helper

### DISCUSSION THREAD

**Discussion by Zorakie** [arnoudkooi/SN-Utils#479] [Ideas]
Hi, this is a long shot, as I know this would need quite a bit of maintenance to keep it alive, but anyway:

What I often miss is, if you have a random field where you can write a piece of script, you don't have available objects anywhere.

Example: Today, I needed to do some prefilling over MRVS and without "how to do it" googling, I would have no idea there is some g_service_catalog API... well, just because I am more of an integration guy and I do not do catalog that often to "just know".

Now, of course, there usually is documentation (but yeah, not always and not for everything), but at least for me, it would be great to see right from the bet, what all I can use. Those APIs are usually very reasonably descriptive, so when you see some API, and when you see its keys/functions, you just easily find what you need without scratching your head, logging stuff just to find out what all is in scope, or just plain googling.

How I could imagine this feature:
A new icon near a field label, this icon pops up some modal with objects.
Those objects would be traversable/openable, basically in a similar way as you would inspect them on the console. Minus various prototypes, cyclic keys/references, and other garbage.

How we could get the data?
Well, this is more tricky. Obviously, SN keeps evolving, and manually adding various objects could be challenging. So, let's automate (partially).
We can collaborate on a simple document, where we identify all the fields that should be in scope.
For each field - and this could be a challenge to achieve - we could have an automated script that logs those objects/sends the result to some repo.
From this repo, SNUtils would update it regularly between releases.

Where to run this? ... Maybe SNDevs instance? Not sure who owns it, but it could be a good place.

So example:
Business rule, script
altered function passing global this, then recursive function to get all keys, subkeys, etc, storing data to table, telling, in BR, this is json of keys.
Client script:
Similar scenario, a script that would get all available APIs and by Ajax calling some REST call to store it to the repo, and then, for example, running some ATF, robot framework, or something, that would open a form with this demo client script.

It may sound a bit clunky and I definitely understand this would be a bigger project, but hell, this is something I miss in SN for the whole of my career.  

Thx :)

**Comment by arnoudkooi** [comment]
Don't understand what you envision, but external calls are not an option.
Besides feel like if it's not well maintained / complete it adds more steps.
